README.md
# Ambassador Auth JWT-RSA Service

This is a fork of [kminehart/ambassador-auth-jwt](https://github.com/kminehart/ambassador-auth-jwt), which is able to verify HMAC based tokens, but not RSA ones. This module is only meant to verify RSA JWTs.

## Using the service

This service is not responsible for creating and assigning JWTs.

It decodes JWT / `Bearer` tokens (provided by the `Authorization` header) and verify the token against a JWKSet, provided by the `JWT_ISSUER` env variable.

It will return a 200 if it can verify the token, 401 if not.

## Configuration

Provide the following environment variables:

| name | description | default value |
|------|-------------|---------------|
| `JWT_ISSUER` | public endpoint with JWKSet (A set of public key) to verify tokens against | |
| `JWT_OUTBOUND_HEADER` | The name of the header to put the decoded payload in | `X-JWT-PAYLOAD` |
| `CHECK_EXP` | check if the token is expired or not | `true` |
| `ALLOW_BASIC_AUTH_PASSTHROUGH` | allow basic auth requests, without a token, to pass through  | `false` |
| `ALLOW_BASIC_AUTH_HEADERS` | comma separated list of headers that could have basic auth credentials  | `Authorization` |
| `ALLOW_BASIC_AUTH_PATH_REGEX` | specify a regex to test the path of the request determine if a basic auth request should be allowed | `^/.*` |

## Run on Kubernetes

A helm chart is included as a git submodule in the helm directory. You can check out the chart at https://github.com/tomwganem/ambassador-auth-jwt-helm

## Logging

All logs are in json format to be consumed in a ELK stack.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.4.4-rc.0](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.4.3...v0.4.4-rc.0) (2021-03-10)


### Features

* **OPSENG-985:** supporting multiple jwt issuers ([5d979d7](https://github.com/tomwganem/ambassador-auth-jwt/commit/5d979d774751b2c13fb020ff7df3589767bb3ac3))

### [0.4.3](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.4.2...v0.4.3) (2021-03-10)

### [0.4.2](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.4.1...v0.4.2) (2021-03-08)


### Features

* distroless ([5b1a804](https://github.com/tomwganem/ambassador-auth-jwt/commit/5b1a8048457d2119bae3b7a6c17f784f3c50c450))

### [0.4.1](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.4.0...v0.4.1) (2020-05-01)

<a name="0.4.0"></a>
# [0.4.0](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.3.0...v0.4.0) (2019-07-23)


### Bug Fixes

* allow basic auth credentials in the Authorization header to pass ([222c287](https://github.com/tomwganem/ambassador-auth-jwt/commit/222c287))
* allow setting log level from environment variable ([c234b3f](https://github.com/tomwganem/ambassador-auth-jwt/commit/c234b3f))
* delete "sub" key from claims ([e3867ba](https://github.com/tomwganem/ambassador-auth-jwt/commit/e3867ba))
* fix basic auth regex check ([7f91aed](https://github.com/tomwganem/ambassador-auth-jwt/commit/7f91aed))
* go mod tidy ([e28e7cf](https://github.com/tomwganem/ambassador-auth-jwt/commit/e28e7cf))
* provide backward compatibility in the error msg structure returned ([0831fa7](https://github.com/tomwganem/ambassador-auth-jwt/commit/0831fa7))
* set CHECK_EXP to true by default ([328763f](https://github.com/tomwganem/ambassador-auth-jwt/commit/328763f))
* update submodule ([7d6015c](https://github.com/tomwganem/ambassador-auth-jwt/commit/7d6015c))


### Features

* allow Basic Auth Requests to be passed through ([f6c5f36](https://github.com/tomwganem/ambassador-auth-jwt/commit/f6c5f36))
* allows passing basic auth creds in multiple headers ([93c6755](https://github.com/tomwganem/ambassador-auth-jwt/commit/93c6755))
* provide more structured error message ([bcab0f1](https://github.com/tomwganem/ambassador-auth-jwt/commit/bcab0f1))
* specify header to allow basic auth pass through ([c409aff](https://github.com/tomwganem/ambassador-auth-jwt/commit/c409aff))
* use golang version 1.12.6 ([8c9c9cb](https://github.com/tomwganem/ambassador-auth-jwt/commit/8c9c9cb))
* use regex to allow basic auth requests ([87bde66](https://github.com/tomwganem/ambassador-auth-jwt/commit/87bde66))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.2.0...v0.3.0) (2019-03-08)


### Features

* add helm chart as sub repo ([b2ddcc2](https://github.com/tomwganem/ambassador-auth-jwt/commit/b2ddcc2))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/tomwganem/ambassador-auth-jwt/compare/v0.1.0...v0.2.0) (2019-03-07)


### Bug Fixes

* fix issue where we are unable to auth tokens with 'expires_at' ([2de4949](https://github.com/tomwganem/ambassador-auth-jwt/commit/2de4949))


### Features

* add sentry ([967e21e](https://github.com/tomwganem/ambassador-auth-jwt/commit/967e21e))



<a name="0.1.0"></a>
# 0.1.0 (2019-02-14)


### Bug Fixes

* fix timecode parse ([3efb577](https://github.com/tomwganem/ambassador-auth-jwt/commit/3efb577))
* improve documentation for httpserver package ([9b43d99](https://github.com/tomwganem/ambassador-auth-jwt/commit/9b43d99))
* log in nanoseconds ([2000b7b](https://github.com/tomwganem/ambassador-auth-jwt/commit/2000b7b))
* use token's KeyID to find public key to verify with ([94e6c3c](https://github.com/tomwganem/ambassador-auth-jwt/commit/94e6c3c))


### Features

* assume that "exp" field will be in unix epoch time instead of rfc3339 ([40498e8](https://github.com/tomwganem/ambassador-auth-jwt/commit/40498e8))
* decode token for url params ([477899b](https://github.com/tomwganem/ambassador-auth-jwt/commit/477899b))
* enable cors ([2444a8f](https://github.com/tomwganem/ambassador-auth-jwt/commit/2444a8f))
* move DecodeHTTPHandler to httpserver package ([c4d0ec8](https://github.com/tomwganem/ambassador-auth-jwt/commit/c4d0ec8))
* verify RSA jwtokens by getting a JWKSet from a url ([57f1441](https://github.com/tomwganem/ambassador-auth-jwt/commit/57f1441))


pkg/token/token_test.go
package token

import (
	"testing"
)

func TestDecode(t *testing.T) {

	// var rsaPrivateKey *rsa.PrivateKey
	// // var signeroptions jose.SignerOptions

	// keyid := time.Now().String()
	// // header := jose.HeaderKey("kid")
	// // signeropts := signeroptions.WithHeader(header, keyid)
	// // signeropts := signeroptions.WithType("JWT")
	// // signerkey := jose.SigningKey{Algorithm: jose.RS256, Key: rsaPrivateKey}
	// // sig, err := jose.NewSigner(signerkey, signeropts)
	// sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: rsaPrivateKey}, nil)
	// // if err != nil {
	// // 	panic(err)
	// // }

	// cl := jwt.Claims{
	// 	Subject:  "admin@example.com",
	// 	Issuer:   "https://localhost/api/v1/oauth2/token",
	// 	Expiry:   jwt.NewNumericDate(time.Now().Add(time.Hour * 24)),
	// 	IssuedAt: jwt.NewNumericDate(time.Now()),
	// 	Audience: jwt.Audience{"leela", "fry"},
	// }
	// builder := jwt.Signed(sig)
	// builder = builder.Claims(cl)

	// fmt.Println(json.Marshal(builder))
	// key := jose.JSONWebKey{
	// 	Algorithm: "RS256",
	// 	Key:       rsaPrivateKey.PublicKey,
	// 	KeyID:     keyid,
	// }
	// keys := []jose.JSONWebKey{key}

	// keyset := jose.JSONWebKeySet{Keys: keys}
	// // jwt := "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjIwMTgtMDYtMDZUMjI6MzU6MTQrMDA6MDAifQ.eyJ1c2VyX2lkIjoiNjY0Iiwic2NvcGUiOiJhZG1pbi11c2VyOmFsbCIsImV4cGlyZXNfYXQiOiIyMDE4LTA3LTIyVDE5OjM5OjAwWiIsInV1aWQiOiJiMTcwYTU5Ny0xMDE1LTQwNjAtOGY1OS0xMzdjNDYzNzNkNzAiLCJvcmdhbml6YXRpb25faWQiOiI1OCIsImV4cCI6MTUzMjI4ODM0MCwic3ViIjoidGdhbmVtQGFzcGVyYXNvZnQuY29tIiwibmFtZSI6IlRvbSBHYW5lbSIsImdpdmVuX25hbWUiOiJUb20iLCJmYW1pbHlfbmFtZSI6IkdhbmVtIiwiaWF0IjoxNTMyMjQ1MTQwLCJpc3MiOiJodHRwczovL2FwaS5pYm1hc3BlcmEuY29tL2FwaS92MS9vYXV0aDIvdG9rZW4iLCJpYm1pZF9pZCI6IklCTWlkLTUwQzRUV05QSDEiLCJpZCI6ImFvYy02NjQiLCJyZWFsbWlkIjoiYW9jLWlibWlkIiwiaWRlbnRpZmllciI6IjY2NCJ9.l49rDjfBp4PKzE176zbkus7_BeizRvsOGw_N85vHgGHLSiHwK07fX81vJbK1CR6vXU57CsrdXqy6cGnnk5mtLzghQY_meDAxB745t5c3Rx5wbJLe11BtiWE3LN4EvAoqCVdL8BP9gIzJE-n3_nrLcgXVJDKBBvpI0H52JHLvI0vJRHfdAaRzm-GKMUeL1q0VSzFLhf30O-hBWi3Edz_SBrbMlQY2LYUaJ_cMXNmvt_YflXnx5md7mYl5BmCbWeSCsnG2aucDB7tUK3_YQuWDglqgzQrKP2grdOdldDeUSIUKsZU_rJIKIEKRiy6hiBGA6zdMCcTGakZj2cPcR4OkCw"

	// // keyjson := `{"keys":[{"kid":"2018-05-23T18:40:16+00:00","alg":"RS256","e":"AQAB","n":"1Tg29euEMj8_dQW3EgYdlhOtDQ277jTVAGcMpn2wXWmHgeWX555T7Ni_N3gPDaRlERaWC7KrZ6oCDV20llZK2B9_uB0VObsnva-39SyRHty4R0iizKHiog0a1htkPsPT7gcZNMmZoWbCdS2T8HQBLUwCRWi0FMsxoCc_M3b9IZJxOU_mwVTuH-wyUovTziLJ8oMaiWoDJVg-veb55ybQyI1unrA4mfaNvK1eXfeXfu8BNpsXh6knpllfw9roCLsVgj4qZVgDyw7MUILn4YTw-sYDsUmjo1DdH34DTkD3nvpTP7BqDZW1w02eZCwsma6rg6FRlPDx6_Rorr4IP22O9Q","kty":"RSA"},{"kid":"2018-06-06T22:35:14+00:00","alg":"RS256","e":"AQAB","n":"3erR0oihAaiS2PXHqFH0vh_fsu_GLBYt05-GsGpSdpkJpTq_7ERH-DdyWqab5Dm_aiP-73TSlYG46vYZMG6AN6y53gWwA5T7cx_FZW950c-VySpG-nwN1mTafM4yefofsQJ4Ues3KP-YaSdeBQqtlrem7x9qWTvoqV-WLiAsp1I8NZEjvj68CxjO32ocDc3o3YtBnZlMMMJeWbxcbD9E6oQNfpiHnzPd2L7IiHfFpf5vcWCnp1epTPRsw9HTBCeQ0mqmhzYoFqGwOxS4F5IPOaS92rHhrmo8nYzH2sEsO3gSJiPdR1jplK9UTei3HFVq4kABNGoHmyJaj5R8KHgnCw","kty":"RSA"}]}`
	// // json.Unmarshal([]byte(keyjson), &keyset)
	// // issuer := "http://localhost/.well-known/jwks.json"
	// // a, b, err := Decode(builder, keyset, issuer)
	// // fmt.Println(a)
	// fmt.Println(keyset)
	// assert.Nil(t, err)
}

func TestJwkSetGet(t *testing.T) {
}


pkg/token/decode_test.go
package token

import (
	"testing"
)

func TestDecode(t *testing.T) {
}


pkg/token/decodehttp_test.go
package token

import (
	"testing"
)

func TestDecodeHttp(t *testing.T) {

}


main.go
package main

import (
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
	"encoding/json"

	raven "github.com/getsentry/raven-go"
	log "github.com/sirupsen/logrus"
	"github.com/tomwganem/ambassador-auth-jwt/pkg/httpserver"
)

var (
	// Version should correspond to a git tag
	Version = "0.4.4-rc.0"
	// ListenPortStr saves the value extracted from the LISTEN_PORT env var
	ListenPortStr string
	// ListenPort saves LISTEN_PORT as an integer
	ListenPort int
	// JwtIssuer is set by the JWT_ISSUER env variable. It saves the url where the JWKeyset is found
	JwtIssuer map[string]string
	// JwtOutboundHeader defaults to X-JWT-PAYLOAD and is returned in the response
	JwtOutboundHeader string
	// CheckExp is a simple flag to check whether tokens are expired
	CheckExp bool
	// AllowBasicAuthPassThrough control whether basic auth requests get rejected or not
	AllowBasicAuthPassThrough bool
	// AllowBasicAuthHeaders specifies the header that the basic auth creds are in
	AllowBasicAuthHeaders string
	// AllowBasicAuthPathRegex specifies the path that basic auth requests are allowed on
	AllowBasicAuthPathRegex string
	// NewErrorMessageRegex specifies the paths that we return the new error structure for (needed for backwards compatibility)
	NewErrorMessageRegex string
)

func init() {
	raven.SetDSN(os.Getenv("SENTRY_DSN"))
	raven.SetEnvironment(os.Getenv("SENTRY_CURRENT_ENV"))
	switch logLevel := os.Getenv("LOG_LEVEL"); logLevel {
	case "TRACE":
		log.SetLevel(log.TraceLevel)
	case "DEBUG":
		log.SetLevel(log.DebugLevel)
	case "INFO":
		log.SetLevel(log.InfoLevel)
	case "WARN":
		log.SetLevel(log.WarnLevel)
	case "ERROR":
		log.SetLevel(log.ErrorLevel)
	case "FATAL":
		log.SetLevel(log.FatalLevel)
	case "PANIC":
		log.SetLevel(log.PanicLevel)
	default:
		log.SetLevel(log.InfoLevel)
	}
	raven.SetRelease(Version)
	log.SetFormatter(&log.JSONFormatter{TimestampFormat: time.RFC3339Nano})
	log.SetOutput(os.Stdout)
	log.WithFields(log.Fields{
		"version":            Version,
		"sentry_dsn":         os.Getenv("SENTRY_DSN"),
		"sentry_environment": os.Getenv("SENTRY_CURRENT_ENV"),
	}).Info("Starting ambassador-auth-jwt")
	ListenPortStr = os.Getenv("LISTEN_PORT")
	var err error
	ListenPort, err = strconv.Atoi(ListenPortStr)
	if err != nil {
		log.Warn("Unable to convert LISTEN_PORT to integer, defaulting to port 3000")
		ListenPort = 3000
	}

	err = json.Unmarshal([]byte(os.Getenv("JWT_ISSUER")), &JwtIssuer)
	if err != nil {
		log.Fatal("Could not parse JWT_ISSUER ")
	}

	//JwtIssuer = tempJwtIssuer.(map[string]string)
	JwtOutboundHeader = os.Getenv("JWT_OUTBOUND_HEADER")
	AllowBasicAuthHeaders := os.Getenv("ALLOW_BASIC_AUTH_HEADERS")
	AllowBasicAuthPathRegex := os.Getenv("ALLOW_BASIC_AUTH_PATH_REGEX")
	NewErrorMessageRegex := os.Getenv("NEW_ERROR_MESSAGE_REGEX")

	checkExp := os.Getenv("CHECK_EXP")
	allowBasicAuthPassThrough := os.Getenv("ALLOW_BASIC_AUTH_PASSTHROUGH")

	CheckExp = true
	if checkExp != "" {
		b, err := strconv.ParseBool(checkExp)
		if err != nil {
			log.Warn("Unable to convert CHECK_EXP to bool: setting to true")
			b = true
		}
		CheckExp = b
	}

	AllowBasicAuthPassThrough = false
	if allowBasicAuthPassThrough != "" {
		b, err := strconv.ParseBool(allowBasicAuthPassThrough)
		if err != nil {
			log.Warn("Unable to convert ALLOW_BASIC_AUTH_PASSTHROUGH to bool: setting to false")
			b = false
		}
		AllowBasicAuthPassThrough = b
	}

	httpserver.JwtIssuer = JwtIssuer
	httpserver.JwtCheckExp = CheckExp
	httpserver.AllowBasicAuthPassThrough = AllowBasicAuthPassThrough
	if AllowBasicAuthHeaders != "" {
		httpserver.AllowBasicAuthHeaders = strings.Split(AllowBasicAuthHeaders, ",")
	}
	if AllowBasicAuthPathRegex != "" {
		httpserver.AllowBasicAuthPathRegex = regexp.MustCompile(AllowBasicAuthPathRegex)
	}
	if NewErrorMessageRegex != "" {
		httpserver.NewErrorMessageRegex = regexp.MustCompile(NewErrorMessageRegex)
	}
	if JwtOutboundHeader != "" {
		httpserver.JwtOutboundHeader = JwtOutboundHeader
	}
}

func main() {
	issuers := make([]string, 0, len(JwtIssuer))
	for _,issuer := range JwtIssuer {
	    issuers = append(issuers, issuer)
	}
	server := httpserver.NewServer(issuers)
	log.Fatal(server.Start(ListenPort))
}


pkg/token/token.go
package token

import (
	"crypto/rsa"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	raven "github.com/getsentry/raven-go"
	log "github.com/sirupsen/logrus"
	"gopkg.in/square/go-jose.v2"
	jwt "gopkg.in/square/go-jose.v2/jwt"
)

// JwkSetGet will call the url provided JWT_ISSUER and retreive a JWK Set.
func JwkSetGet(issuer string) (jose.JSONWebKeySet, error) {
	keyset := jose.JSONWebKeySet{}
	resp, err := http.Get(issuer)
	if err != nil {
		return keyset, err
	}
	defer resp.Body.Close()
	if err := json.NewDecoder(resp.Body).Decode(&keyset); err != nil && err != io.EOF {
		return keyset, err
	}
	log.WithFields(log.Fields{
		"keyset": keyset,
		"issuer": issuer,
	}).Info("Retreiving Keyset")
	return keyset, nil
}

// Initialize issuer with its jwk set
func JwkSetGetMap(issuers []string) (map[string]jose.JSONWebKeySet, error) {
	keysetIssuerMap := make(map[string]jose.JSONWebKeySet)
	for _, issuer := range issuers {
		keyset := jose.JSONWebKeySet{}
		resp, err := http.Get(issuer)
		if err != nil {
			return keysetIssuerMap, err
		}
		defer resp.Body.Close()
		if err := json.NewDecoder(resp.Body).Decode(&keyset); err != nil && err != io.EOF {
			return keysetIssuerMap, err
		}
		log.WithFields(log.Fields{
			"keyset": keyset,
			"issuer": issuer,
		}).Info("Retreiving Keyset")
		keysetIssuerMap[issuer] = keyset
	}
	return keysetIssuerMap, nil
}


// Decode the raw token and validate it with a JWK Set.
func Decode(jwtoken string, jwkset jose.JSONWebKeySet, issuer string) (map[string]interface{}, jose.JSONWebKeySet, error) {
	claims := struct {
		*jwt.Claims
		ExpiresAt      string `json:"expires_at,omitempty"`
		Scope          string `json:"scope,omitempty"`
		OrganizationID string `json:"organization_id,omitempty"`
		UUID           string `json:"uuid,omitempty"`
	}{}
	mapClaims := make(map[string]interface{})
	token, err := jwt.ParseSigned(jwtoken)
	if err != nil {
		return mapClaims, jwkset, fmt.Errorf("Could not read jwt")
	}
	keyid := token.Headers[0].KeyID
	jwk := jwkset.Key(keyid)
	if len(jwk) == 0 {
		jwkset, err = JwkSetGet(issuer)
		if err != nil {
			raven.CaptureErrorAndWait(err, nil)
			log.WithFields(log.Fields{
				"issuer": issuer,
				"err":    err,
			}).Fatal("Unable to update keyset")
		}
		log.WithFields(log.Fields{
			"keyset": jwkset,
			"issuer": issuer,
		}).Info("Updating Keyset")
		jwk = jwkset.Key(keyid)
		if len(jwk) == 0 {
			return mapClaims, jwkset, errors.New("Can not find token's key id in jwk set")
		}
	}

	if err := token.Claims(jwk[0].Key.(*rsa.PublicKey), &claims); err != nil {
		raven.CaptureError(err, nil)
		return mapClaims, jwkset, err
	}
	marshalClaims, err := json.Marshal(claims)
	if err != nil {
		raven.CaptureError(err, nil)
		return mapClaims, jwkset, err
	}
	if err := json.Unmarshal(marshalClaims, &mapClaims); err != nil {
		raven.CaptureError(err, nil)
		return mapClaims, jwkset, err
	}

	return mapClaims, jwkset, nil
}


pkg/token/decode.go
package token

import (
	"crypto/rsa"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	log "github.com/sirupsen/logrus"
	"gopkg.in/square/go-jose.v2"
	jwt "gopkg.in/square/go-jose.v2/jwt"
)

// JwkSetGet will call the url provided JWT_ISSUER and retreive a JWK Set.
func JwkSetGet(issuer string) (jose.JSONWebKeySet, error) {
	keyset := jose.JSONWebKeySet{}
	resp, err := http.Get(issuer)
	if err != nil {
		return keyset, err
	}
	defer resp.Body.Close()
	if err := json.NewDecoder(resp.Body).Decode(&keyset); err != nil && err != io.EOF {
		return keyset, err
	}
	log.WithFields(log.Fields{
		"keyset": keyset,
		"issuer": issuer,
	}).Info("Retreiving Keyset")
	return keyset, nil
}

// Decode the raw token and validate it with a JWK Set.
func Decode(jwtoken string, jwkset jose.JSONWebKeySet, issuer string) (jwt.Claims, jose.JSONWebKeySet, error) {
	claims := jwt.Claims{}
	token, err := jwt.ParseSigned(jwtoken)
	if err != nil {
		return claims, jwkset, fmt.Errorf("Could not read jwt")
	}
	keyid := token.Headers[0].KeyID
	jwk := jwkset.Key(keyid)
	if len(jwk) == 0 {
		jwkset, err = JwkSetGet(issuer)
		if err != nil {
			log.WithFields(log.Fields{
				"issuer": issuer,
			}).Fatal("Unable to update keyset")
		}
		log.WithFields(log.Fields{
			"keyset": jwkset,
			"issuer": issuer,
		}).Info("Updating Keyset")
		jwk = jwkset.Key(keyid)
		if len(jwk) == 0 {
			return claims, jwkset, errors.New("Can not find token's key id in jwk set")
		}
	}

	if err := token.Claims(jwk[0].Key.(*rsa.PublicKey), &claims); err != nil {
		return claims, jwkset, err
	}

	return claims, jwkset, nil
}


pkg/httpserver/server.go
package httpserver

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	raven "github.com/getsentry/raven-go"
	log "github.com/sirupsen/logrus"
	"github.com/tomwganem/ambassador-auth-jwt/pkg/token"
	"gopkg.in/square/go-jose.v2"
)

// Error is returned in ErrorMsg
type Error struct {
	Code    string `json:"code,omitempty"`
	Message string `json:"message,omitempty"`
}

// ErrorMsg is returned on unauthorized requests
type ErrorMsg struct {
	StatusCode int     `json:"status_code,omitempty"`
	Errors     []Error `json:"errors"`
}

var (
	// JwtCheckExp will determine if we need to verify if the token is expired or not
	JwtCheckExp = true
	// JwtIssuer is the url where we can retrieve a set of public keys to verify rsa based tokens with
	JwtIssuer = map[string]string{
		"default": "http://localhost/.well-known/jwks.json",
	}
	// JwtOutboundHeader is the name of header the parsed token claims will be inserted into
	JwtOutboundHeader = "X-JWT-PAYLOAD"
	// AllowBasicAuthPassThrough will allow requests with a basic auth authorization header to be passed through
	AllowBasicAuthPassThrough = false
	// AllowBasicAuthHeaders specifies the header to extract the basic auth request from
	AllowBasicAuthHeaders = []string{"Authorization"}
	// AllowBasicAuthPathRegex allows us to only allow a basic auth pass through for requests with a certain path
	AllowBasicAuthPathRegex = regexp.MustCompile(`^\/.*`)
	// NewErrorMessageRegex is used to apply new error structure only to specific endpoints (see bcab0f12220b1e26c6dad10ee6aa1825172051ab). This is for backward compatibility.
	NewErrorMessageRegex = regexp.MustCompile(`^\/.*`)
	// BasicAuthRegex is for checking if a basic auth request is formatted correctly
	BasicAuthRegex = regexp.MustCompile(`^Basic\ .*`)
)

// Server needs to know about the Issuer url to verify tokens against
type Server struct {
	IssuerJwkSetMap map[string]jose.JSONWebKeySet
}

// Start accepting requests and decoding Authorization headers
func (server *Server) Start(port int) error {
	// jwks, err := token.JwkSetGet(server.Issuer)
	http.HandleFunc("/", server.DecodeHTTPHandler)
	return http.ListenAndServe(fmt.Sprintf("0.0.0.0:%d", port), nil)
}

// DecodeHTTPHandler will try to extract the bearer token found in the Authorization header of each request and verify it
func (server *Server) DecodeHTTPHandler(w http.ResponseWriter, r *http.Request) {
	q, _ := url.ParseQuery(r.URL.RawQuery)
	successFields := log.Fields{
		"remote_addr": r.RemoteAddr,
		"host":        r.Host,
		"method":      r.Method,
		"path":        r.URL.Path,
		"query":       q,
		"user_agent":  r.UserAgent(),
		"status":      "200",
	}
	errorFields := log.Fields{
		"remote_addr": r.RemoteAddr,
		"host":        r.Host,
		"method":      r.Method,
		"path":        r.URL.Path,
		"query":       q,
		"user_agent":  r.UserAgent(),
		"status":      "401",
	}
	debugFields := log.Fields{
		"method": r.Method,
		"path":   r.URL.Path,
		"query":  q,
	}
	successLogger := log.WithFields(successFields)
	errorLogger := log.WithFields(errorFields)
	debugLogger := log.WithFields(debugFields)

	unauthorizedOld := map[string]string{"code": "unauthorized", "message": "You are not authorized to perform the requested action"}

	unauthorizedNew := ErrorMsg{
		StatusCode: 401,
		Errors: []Error{
			{
				Code:    "unauthorized",
				Message: "You are not authorized to perform the requested action",
			},
		},
	}

	matchedPath := NewErrorMessageRegex.Match([]byte(r.URL.Path))
	var error []byte
	if matchedPath {
		error, _ = json.Marshal(unauthorizedNew)
	} else {
		error, _ = json.Marshal(unauthorizedOld)
	}

	enableCors(&w)
	// Enabled PREFLIGHT calls
	if r.Method == "OPTIONS" {
		successLogger.Info("CORS Request OK")
		return
	}

	auth := r.Header.Get("Authorization")
	query := r.URL.Query()
	t := query["token"]
	bt := query["bearer_token"]
	basicAuthAllowed, msg := basicAuthPassCheck(r, debugLogger)
	matchedAuth := BasicAuthRegex.Match([]byte(auth))
	// Allows basic auth credentials in the Authorization header to be passed through
	if matchedAuth && basicAuthAllowed {
		successLogger.Info(msg)
		return
	}
	// The following checks for tokens passed as a query parameter
	if auth == "" && !basicAuthAllowed {
		if len(t) < 1 || t[0] == "" {
			if len(bt) < 1 || bt[0] == "" {
				errorLogger.Warn("Unable to retrieve JWToken from Authorization header or query parameter. " + msg)
				http.Error(w, string(error), 401)
				return
			}
			auth = bt[0]
		} else {
			auth = t[0]
		}

	} else if auth == "" && basicAuthAllowed {
		log.WithFields(successFields).Info(msg)
		return
	}

	claims := make(map[string]interface{})
	auth = strings.Replace(auth, "Bearer ", "", 1)
	found,issuer := getJwtIssuer(r)
	if !found {
		errorLogger.Error("Could not find jwt issuer for path " + r.URL.Path)
		http.Error(w, string(error), 401)
		return
	}
	claims, jwkset, err := token.Decode(auth, server.IssuerJwkSetMap[issuer], issuer)
	server.IssuerJwkSetMap[issuer] = jwkset
	if err != nil {
		errorLogger.Error(err.Error())
		http.Error(w, string(error), 401)
		return
	}
	exp := time.Now()
	if JwtCheckExp {
		// Checks to see if the there is an "exp" field
		if _, ok := claims["exp"]; ok != true {
			// Checks to see if there is an "expires_at" field. Note: "expires_at" doesn't follow the RFC and shouldn't be a field in most JWTokens. It's the same as "exp", except it's in RFC3339.
			if _, ok := claims["expires_at"]; ok != true {
				errorLogger.Error(err.Error())
				http.Error(w, string(error), 401)
				return
			}
			exp, err = time.Parse(time.RFC3339, claims["expires_at"].(string))
			if err != nil {
				raven.CaptureError(err, nil)
				errorLogger.Error(err.Error())
				http.Error(w, string(error), 500)
			}
		} else {
			exp = time.Unix(int64(claims["exp"].(float64)), 0)
		}

		now := time.Now()

		if exp.Before(now) {
			errorLogger.Error("Token is expired")
			http.Error(w, string(error), 401)
			return
		}
	}
	marshaledClaims, err := json.Marshal(claims)
	successFields["claims"] = claims
	log.WithFields(successFields).Info("Authentication Success")
	w.Header().Set(JwtOutboundHeader, string(marshaledClaims))
}

// NewServer creates a new Server object with the jwkset retrieved from the issuer
func NewServer(issuers []string) *Server {
	jwks, err := token.JwkSetGetMap(issuers)
	if err != nil {
		raven.CaptureErrorAndWait(err, nil)
		log.WithFields(log.Fields{
			"issuerJwkSetMap": jwks,
			"err":    err,
		}).Fatal("Unable to retrieve keyset")
	}

	return &Server{
		IssuerJwkSetMap: jwks,
	}
}

// enableCors sets some hardcoded headers for OPTIONS requests. We return all OPTIONS requests with a 200.
func enableCors(w *http.ResponseWriter) {
	(*w).Header().Set("Access-Control-Allow-Origin", "*")
	(*w).Header().Set("Access-Control-Allow-Methods", "GET, POST, DELETE, PUT, OPTIONS")
	(*w).Header().Set("Access-Control-Allow-Headers", "authorization")
	(*w).Header().Set("Access-Control-Max-Age", "1728000")
	(*w).Header().Set("Access-Control-Expose-Headers", "")
}

// basicAuthPassCheck returns a boolean. It will return true if:
// 1. ALLOW_BASIC_AUTH_PASSTHROUGH is set to true
// 2. the path of the request matches ALLOW_BASIC_AUTH_PATH_REGEX
// 3. the value passed in ALLOW_BASIC_AUTH_HEADERS looks like a valid basic auth value (i.e. starts with "Basic", can be base64 decoded, can be split into a username:password pair)
func basicAuthPassCheck(r *http.Request, debugLogger *log.Entry) (bool, string) {
	debugLogger.Trace(fmt.Sprintf("ALLOW_BASIC_AUTH_PASSTHROUGH set to %t", AllowBasicAuthPassThrough))
	if AllowBasicAuthPassThrough {
		debugLogger.Trace(fmt.Sprintf("ALLOW_BASIC_AUTH_HEADERS have values: %v", AllowBasicAuthHeaders))
		for _, header := range AllowBasicAuthHeaders {
			b, msg := basicAuthHeaderCheck(header, r, debugLogger)
			if b {
				return b, msg
			}
		}
	}
	return false, "Basic Auth Not Allowed"
}

func basicAuthHeaderCheck(header string, r *http.Request, debugLogger *log.Entry) (bool, string) {
	debugLogger.Trace(fmt.Sprintf("Checking value in header: %s", header))
	path := r.URL.Path
	debugLogger.Trace(fmt.Sprintf("ALLOW_BASIC_AUTH_PATH_REGEX set to: %s", AllowBasicAuthPathRegex))
	matchedPath := AllowBasicAuthPathRegex.Match([]byte(path))
	basicAuth := r.Header.Get(header)
	matchedAuth := BasicAuthRegex.Match([]byte(basicAuth))
	if matchedAuth {
		debugLogger.Trace(fmt.Sprintf("header: %s, does have a value that includes: %s", header, BasicAuthRegex))
		if matchedPath {
			debugLogger.Trace(fmt.Sprintf("request path: %s, correctly matches regex: %s", path, AllowBasicAuthPathRegex))
			basicAuth = strings.Replace(basicAuth, "Basic ", "", 1)
			payload, err := base64.StdEncoding.DecodeString(basicAuth)
			if err != nil {
				debugLogger.Trace(fmt.Sprintf("basic auth value: %s can not be base64 decoded", basicAuth))
				return false, "Basic Auth Not Allowed"
			}
			pair := strings.SplitN(string(payload), ":", 2)
			if len(pair) == 2 {
				return true, "Basic Auth Allowed"
			}
			debugLogger.Trace(fmt.Sprintf("decoded basic auth value: %s is unable to be split into a username/password pair", payload))
		} else {
			debugLogger.Trace(fmt.Sprintf("request path: %s, does not match regex: %s", path, AllowBasicAuthPathRegex))
		}
	} else {
		debugLogger.Trace(fmt.Sprintf("header: %s, does not have value that matches: %s", header, BasicAuthRegex))
	}
	return false, "Basic Auth Not Allowed"
}

func getJwtIssuer(r *http.Request) (bool,string) {
	 path := r.URL.Path
	 for jwt_path, issuer := range JwtIssuer {
		 if (strings.Contains(path, jwt_path)) {
			 return true,issuer
		 }
	 }
	 return false,"Path not found in jwt keys"
}
